rlmolecule/mcts/mcts.py

```diff
@@ -30,59 +33,58 @@ def problem(self) -> MCTSProblem:
+        :param state: the starting state, or if not provided, the state returned from _get_root()
+        :param action_selection_function: a function used to select among the possible next actions. Defaults to
+            softmax sampling by visit counts.
+        :return: The search path (as a list of vertexes) and the final reward from the last state.
+        :param max_depth: the maximum number of times to recurse. Defaults to system recursion limit.
+        """
         vertex = self._get_root() if state is None else self.get_vertex_for_state(state)
-        action_selection_function = self.softmax_selection if explore else self.visit_selection
-        num_mcts_samples = self._num_mcts_samples if num_mcts_samples is None else num_mcts_samples
-
-        def ucb_selection(parent):
-            return max(parent.children, key=lambda child: self._ucb_score(parent, child))
-
-        mcts_selection_function = ucb_selection if mcts_selection_function is None else mcts_selection_function
+        if action_selection_function is None:
+            action_selection_function = self.softmax_selection
 
-        return self.run_from_vertex(vertex, action_selection_function, num_mcts_samples, mcts_selection_function)
-
-    def run_from_vertex(
-            self,
-            vertex: MCTSVertex,
-            action_selection_function: Callable[[MCTSVertex], MCTSVertex],
-            num_mcts_samples: int,
-            mcts_selection_function: Callable[[MCTSVertex], MCTSVertex],
-    ) -> []:
         path: [] = []
-        while True:
-            # path.append(vertex)
-            self.sample(vertex, num_mcts_samples, mcts_selection_function)
+        value = 0.
+        for depth in itertools.count():
+            for _ in range(num_mcts_samples):
+                value = self.sample_from(vertex)
             self._accumulate_path_data(vertex, path)
             children = vertex.children
             if children is None or len(children) == 0:
                 break
+            if depth > (max_depth if max_depth is not None else sys.getrecursionlimit()):
+                logger.warning(f"{self} reached max_depth or recursion limit")
+                break
             vertex = action_selection_function(vertex)
 
-        return path
+        return path, value
 
-    def sample(
-            self,
-            vertex: MCTSVertex,
-            num_mcts_samples: int,
-            mcts_selection_function: Callable[[MCTSVertex], MCTSVertex],
-    ) -> None:
-        for _ in range(num_mcts_samples):
-            search_path, value = self._select(vertex, mcts_selection_function)
-            self._backpropagate(search_path, value)
+    def sample_from(self, vertex: MCTSVertex) -> float:
+        """Run a single MCTS sample from the given vertex
+
+        :param vertex: The starting vertex for MCTS sampling
+        """
+        search_path = self._select(vertex)
+        value = self._evaluate(search_path[-1], search_path)
```

Comment:
Charles, let me know if you're OK with this change or if there's something I'm missing. I figured it might be cleaner to have evaluate live outside select.

Reply:
I agree-- I moved evaluate outside of select, and I kept it with a optional num_mcts_samples parameter. If you only want one sample, that is the default behavior. But, we always do multiple samples, so it is less redundant and simpler if the loop is inside the sample method.

 I'm ok with renaming this "sample_from" if you like, it was just easier to merge without renaming.